src/integrationTest/java/org/opensearch/security/api/RolesRestApiIntegrationTest.java

```diff
@@ -0,0 +1,462 @@
+                return builder.endObject();
+            })))),
+            Map.of("cluster_permissions", "Array expected", "index_permissions", "Array expected")
+        );
+        assertWrongDataType(
+            badRequest(() -> client.patch(apiPath(predefinedRoleName), patch(replaceOp("cluster_permissions", "true")))),
+            Map.of("cluster_permissions", "Array expected")
+        );
+        assertNullValuesInArray(
+            client.patch(
+                apiPath(),
+                patch(
+                    addOp(
+                        randomAsciiAlphanumOfLength(5),
+                        role(randomClusterPermissions(true), randomIndexPermissions(true), randomTenantPermissions(true))
+                    )
+                )
+            )
+        );
+        // TODO related to issue #4426
+        assertNullValuesInArray(
+            client.patch(apiPath(predefinedRoleName), patch(replaceOp("cluster_permissions", randomClusterPermissions(true))))
+        );
+    }
+
+    @Override
+    void forbiddenToCreateEntityWithRestAdminPermissions(final TestRestClient client) throws Exception {
+        forbidden(() -> client.putJson(apiPath("new_rest_admin_role"), roleWithClusterPermissions(randomRestAdminPermission())));
+        forbidden(
+            () -> client.patch(
+                apiPath(),
+                patch(addOp("new_rest_admin_action_group", roleWithClusterPermissions(randomRestAdminPermission())))
+            )
+        );
+    }
+
+    @Override
+    void forbiddenToUpdateAndDeleteExistingEntityWithRestAdminPermissions(final TestRestClient client) throws Exception {
+        // update
+        forbidden(
+            () -> client.putJson(
+                apiPath(REST_ADMIN_PERMISSION_ROLE),
+                role(randomClusterPermissions(false), randomIndexPermissions(false), randomTenantPermissions(false))
+            )
+        );
+        forbidden(
+            () -> client.patch(
+                apiPath(),
+                patch(
+                    replaceOp(
+                        REST_ADMIN_PERMISSION_ROLE,
+                        role(randomClusterPermissions(false), randomIndexPermissions(false), randomTenantPermissions(false))
+                    )
+                )
+            )
+        );
+        forbidden(
+            () -> client.patch(
+                apiPath(REST_ADMIN_PERMISSION_ROLE),
+                patch(replaceOp("cluster_permissions", randomClusterPermissions(false)))
+            )
+        );
+        // remove
+        forbidden(() -> client.patch(apiPath(), patch(removeOp(REST_ADMIN_PERMISSION_ROLE))));
+        forbidden(() -> client.patch(apiPath(REST_ADMIN_PERMISSION_ROLE), patch(removeOp("cluster_permissions"))));
+        forbidden(() -> client.delete(apiPath(REST_ADMIN_PERMISSION_ROLE)));
+    }
+
+    void assertRole(
+        final TestRestClient.HttpResponse response,
+        final String roleName,
+        final Boolean hidden,
```

Comment:
nit: why not pass `hidden` and `reserved` as boolean to avoid the null check below?

Reply:
True I do not like it as well. The main reason why I did it is that the end users can pass `hidden` and `reserved` properties only they use super-admin creds. Whdyt does it make sense to craete an issue? After I've finished, I'm planning to create RFC with the list of problems we have in our REST API.

Reply:
Oh I see now, its because these are passed into `verifyCrudOperations` and ultimately used in `static ToXContentObject role(...)`. Marking this as resolved.